examples/pytorch/adaptive_sampling/adaptive_sampling.py

```diff
@@ -0,0 +1,475 @@
+        # self.fc_self = nn.Linear(in_feats, out_feats, bias=bias).double()
+        self.fc_neigh = nn.Linear(in_feats, out_feats, bias=bias).double()
+        self.reset_parameters()
+        self.G = G
+
+    def reset_parameters(self):
+        gain = nn.init.calculate_gain('relu')
+        # nn.init.xavier_uniform_(self.fc_self.weight, gain=gain)
+        nn.init.xavier_uniform_(self.fc_neigh.weight, gain=gain)
+
+    def forward(self, graph, hidden_feat, node_feat, layer_id, sample_weights, norm_adj=None, var_loss=None,
+                is_test=False):
+        """
+        graph: Bipartite.  Has two edge types.  The first one represents the connection to
+        the desired nodes from neighbors.  The second one represents the computation
+        dependence of the desired nodes themselves.
+        :type graph: dgl.DGLHeteroGraph
+        """
+        # local_var is not implemented for heterograph
+        # graph = graph.local_var()
+
+        neighbor_etype_name = 'block%d' % layer_id
+        src_name = 'layer%d' % layer_id
+        dst_name = 'layer%d' % (layer_id + 1)
+
+        graph.nodes[src_name].data['hidden_feat'] = hidden_feat
+        graph.nodes[src_name].data['node_feat'] = node_feat[graph.layer_mappings[layer_id]]
+        ##normalized degree for node_i is defined as  1/sqrt(d_i+1)
+        ##use the training graph during training and whole graph during testing
+        if not is_test:
+            graph.nodes[src_name].data['norm_deg'] = 1 / torch.sqrt(
+                trainG.in_degrees(graph.layer_mappings[layer_id]).double() + 1)
+            graph.nodes[dst_name].data['norm_deg'] = 1 / torch.sqrt(
+                trainG.in_degrees(graph.layer_mappings[layer_id + 1]).double() + 1)
+        else:
+            graph.nodes[src_name].data['norm_deg'] = 1 / torch.sqrt(
+                allG.in_degrees(graph.layer_mappings[layer_id]).double() + 1)
+            graph.nodes[dst_name].data['norm_deg'] = 1 / torch.sqrt(
+                allG.in_degrees(graph.layer_mappings[layer_id + 1]).double() + 1)
+        graph.nodes[dst_name].data['node_feat'] = node_feat[graph.layer_mappings[layer_id + 1]]
+        graph.nodes[src_name].data['q_probs'] = graph.block_aux_data[layer_id]
+
+        def send_func(edges):
+
+            hu = torch.matmul(edges.src['node_feat'], sample_weights[:, 0])
+            hv = torch.matmul(edges.dst['node_feat'], sample_weights[:, 1])
+            ##attention coeffient is adjusted by normalized degree and sampling probability
+            attentions = edges.src['norm_deg'] * edges.dst['norm_deg'] * (F.relu(hu + hv) + 0.1) / edges.src[
+                'q_probs'] / len(hu)
+
+            hidden = edges.src['hidden_feat'] * torch.reshape(attentions, [-1, 1])
+            return {"hidden": hidden}
+
+        def recv_func(nodes):
+            # print(nodes.mailbox['hidden'].shape)
+            msgs = torch.sum(nodes.mailbox['hidden'], dim=1)
+            return {'neigh': msgs}
```

Comment:
Use `recv_func = dgl.function.sum('hidden', 'neigh')`; this should speed up and shorten your code.